### PR TITLE
Enable BFloat support for gemms on arch other than ampere

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes #{issue number}

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
This should be ready. Test failures are unrelated.

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 1]
```diff
@@ -550,37 +549,26 @@ void gemm<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16)) {
...
-        ldc,
-        CUDA_R_32F,
-        CUBLAS_GEMM_DFALT_TENSOR_OP));
-    // On CUDA versions prior to 11, users are required to set the math mode to CUBLAS_TENSOR_OP_MATH
-    // manually to be able to use tensor cores for FP16. On CUDA 11, this is no longer required.
-    TORCH_CUDABLAS_CHECK(cublasSetMathMode(handle, CUBLAS_DEFAULT_MATH));
-  } else {
-    TORCH_CHECK(false, "BFloat16 gemm in CUDA requires Ampere or later GPU");
-  }
+  TORCH_CUDABLAS_CHECK(cublasGemmEx(
```
Comment (ngimel) [inline aten/src/ATen/cuda/CUDABlas.cpp:552]: setting and resetting cublas MathMode is not required if you specify CUBLAS_GEMM_DFALT_TENSOR_OP?

File: test/test_linalg.py [hunk 2]
```diff
@@ -4501,7 +4498,7 @@ def _test_addmm_addmv(self, f, t, m, v, *, alpha=None, beta=None, transpose_out=
                         torch.cfloat: 1e-4, torch.cdouble: 1e-8})
     @dtypesIfCUDA(*torch.testing.get_all_complex_dtypes(),
                   *([torch.float32, torch.float64, torch.bfloat16]
-                    if TEST_WITH_ROCM else torch.testing.get_all_fp_dtypes(include_bfloat16=AMPERE_OR_ROCM)))
+                    if TEST_WITH_ROCM else torch.testing.get_all_fp_dtypes(include_bfloat16=(CUDA11 and SM53))))
```
Comment (ngimel) [inline test/test_linalg.py:4501]: can you do `include_bfloat16=(TEST_WITH_ROCM or (CUDA11 and SM53))` as in other places here?

File: test/test_linalg.py [hunk 3]
```diff
@@ -4746,15 +4743,20 @@ def test_bmm(self, device, dtype):
...
+        is_cuda_bfloat = False
+        if dtype == torch.bfloat16 and self.device_type == 'cuda':
+            is_supported = TEST_WITH_ROCM or (CUDA11 and SM53)
+            is_cuda_bfloat = True
 
         if not is_supported:
-            b1 = torch.randn(num_batches, M, N, device=device).to(dtype)
-            b2 = torch.randn(num_batches, N, O, device=device).to(dtype)
-            self.assertRaisesRegex(RuntimeError, "type|Type|not implemented|Ampere", lambda: torch.bmm(b1, b2))
+            if not is_cuda_bfloat:
```
Comment (ngimel) [inline test/test_linalg.py:4753]: `is_supported=False`, `is_cuda_bfloat=False` is an impossible situation?

File: torch/testing/_internal/common_cuda.py [hunk 4]
```diff
@@ -15,6 +15,10 @@
 TEST_CUDNN = TEST_CUDA and torch.backends.cudnn.is_acceptable(torch.tensor(1., device=CUDA_DEVICE))
 TEST_CUDNN_VERSION = torch.backends.cudnn.version() if TEST_CUDNN else 0
 
+CUDA11 = torch.version.cuda and float(torch.version.cuda) >= 11
```
Comment (ngimel) [inline torch/testing/_internal/common_cuda.py:18]: These are confusing macros - CUDA11 and SM53 use >= comparisons, and CUDA9 uses `==`

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 5]
```diff
@@ -343,7 +342,7 @@ void bgemm<at::BFloat16>(CUDABLAS_BGEMM_ARGTYPES(at::BFloat16)) {
                                    (int) num_batches, rocblas_datatype_f32_r, rocblas_gemm_algo_standard,
                                    0, 0, NULL, NULL));
   #else
-    TORCH_CHECK(false, "BFloat16 bgemm in CUDA requires Ampere or later GPU");
+    TORCH_CHECK(false, "BFloat16 requires CUDA 11");
```
Comment (mruberry) [inline aten/src/ATen/cuda/CUDABlas.cpp:345]: I suggest adding the function reference back to clarify where this error comes from:

"CUDA bfloat16 bgemm requires CUDA 11 or later"

File: test/test_linalg.py (hunk 3 above)
Comment (mruberry) [inline test/test_linalg.py:4753]: These conditions look collapsible for readability:

```
if self.device_type == 'cuda' and dtype is torch.bfloat16 and not (TEST_WITH_ROCM or (CUDA11 and SM53)):
  # comment and test for an exception or early exit
else:
  # runs test as normal
```

File: test/test_linalg.py [hunk 6]
```diff
@@ -4920,18 +4922,24 @@ def test_addbmm(self, device, dtype):
...
+            else:
+                is_cuda_bfloat = True
+                is_supported = TEST_WITH_ROCM or (CUDA11 and SM53)
 
         if not is_supported:
-            b1 = make_tensor((num_batches, M, N), device, dtype, low=-1, high=1)
-            b2 = make_tensor((num_batches, N, O), device, dtype, low=-1, high=1)
-            t = make_tensor((M, O), device, dtype, low=-1, high=1)
-            self.assertRaisesRegex(RuntimeError, "type|Type|not implemented|Ampere", lambda: torch.addbmm(t, b1, b2))
+            if not is_cuda_bfloat:
```
Comment (mruberry) [inline test/test_linalg.py:4935]: See above suggestion on collapsing this for readability by testing the exact condition on one line.

File: test/test_linalg.py [hunk 7]
```diff
@@ -4986,16 +4994,21 @@ def test_baddbmm(self, device, dtype):
...
+        if dtype == torch.bfloat16 and self.device_type == 'cuda':
+            is_supported = TEST_WITH_ROCM or (CUDA11 and SM53)
+            is_cuda_bfloat = True
 
         if not is_supported:
-            b1 = make_tensor((num_batches, M, N), device, dtype, low=-1, high=1)
-            b2 = make_tensor((num_batches, N, O), device, dtype, low=-1, high=1)
-            t = make_tensor((num_batches, M, O), device, dtype, low=-1, high=1)
-            self.assertRaisesRegex(RuntimeError, "type|Type|not implemented|Ampere", lambda: torch.baddbmm(t, b1, b2))
+            if not is_cuda_bfloat:
```
Comment (mruberry) [inline test/test_linalg.py:5004]: Same collapsing here, too.

File: torch/testing/_internal/common_cuda.py (hunk 4 above)
Comment (mruberry) [reply torch/testing/_internal/common_cuda.py:18]: The CUDA11 and SM53 variables need names like "AtLeastCUDA11" and "AtLeastSM53" or "CUDA11OrLater" and "SM53OrLater" to be clear.

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/cuda/CUDABlas.cpp:552]: According to https://docs.nvidia.com/cuda/cublas/index.html#cublasmath_t
```
CUBLAS_DEFAULT_MATH    This is the default and highest-performance mode that uses compute and intermediate storage precisions with at least the same number of mantissa and exponent bits as requested. Tensor Cores will be used whenever possible.
CUBLAS_TENSOR_OP_MATH    This mode is deprecated and will be removed in a future release. Allows the library to use Tensor Core operations whenever possible. For single precision GEMM routines cuBLAS will use the CUBLAS_COMPUTE_32F_FAST_16F compute type.
```

File: test/test_linalg.py (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_linalg.py:4753]: Some ops are supported on SM52, and some are not. I don't think it worth the maintenance effort to get a clear list on which is supported on which SM. So what I implemented here is:
```
SM >= 53 ---> supported
SM < 53 ---> undefined behavior
```

File: test/test_linalg.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_linalg.py]: fixed

File: torch/testing/_internal/common_cuda.py (hunk 4 above)
Comment (zasdfgbnm) [reply torch/testing/_internal/common_cuda.py]: fixed

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply aten/src/ATen/cuda/CUDABlas.cpp]: fixed

File: test/test_linalg.py (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_linalg.py]: fixed

File: test/test_linalg.py (hunk 6 above)
Comment (zasdfgbnm) [reply test/test_linalg.py]: fixed

File: test/test_linalg.py (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_linalg.py]: fixed

**Comment by zasdfgbnm** [comment]
@ngimel @mruberry I think I have resolved all review comments, and all tests pass.

**Review by mruberry — APPROVED** [review]
Cool! Thanks @zasdfgbnm!

Would you just rebase this? Sorry PyTorch is especially popular these days.

**Comment by zasdfgbnm** [comment]
@mruberry rebased

**Comment by mruberry** [comment]
Internal builds are failing with:

```
    from torch.testing._internal.common_cuda import _get_torch_cuda_version
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/caffe2/fb/high_perf_models/pytorch/torchscript/test/test_ir_bench#binary,link-tree/torch/testing/_internal/common_cuda.py", line 18, in <module>
    CUDA11OrLater = torch.version.cuda and float(torch.version.cuda) >= 11
ValueError: could not convert string to float: '9.2.0'
```

We typically use LooseVersion for version comparisons. See

https://github.com/pytorch/pytorch/blob/869081961824dfb10ee944145c3a8d435571c53b/torch/testing/_internal/common_methods_invocations.py#L1800

for an example.

**Comment by zasdfgbnm** [comment]
@mruberry fixed

**Pull request closed without merging**